### ISSUE TITLE
Remove SelfInitializedFakes from the API

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#798]
The feature
1. is probably not used much
2. is [barely documented](http://fakeiteasy.readthedocs.io/en/stable/creating-fakes/#explicit-creation-options)
3. supports a very restrictive use case - the fake matches calls only on signature
4. can't record calls that throw exceptions
5. requires serialization of method info (not too hard) and out/return values (which may not be available in .NET Core, depending on the types)
6. is [causing angst](https://github.com/FakeItEasy/FakeItEasy/pull/617/files/9ea652d2bc590f081f954c9974aff6171111ef74#r69927243) over in #617.

Ditch it?

**Comment by adamralph** [comment]
I only found one use of it on GitHub, and it looks like an abandoned library: https://github.com/omnibs/service-ghost

**Comment by thomaslevesque** [comment]
:+1:

**Comment by blairconrad** [comment]
Indeed. Of course, public GitHub repos are not the world.
Still.

**Comment by blairconrad** [comment]
Also, you're [redoing my research](https://github.com/FakeItEasy/FakeItEasy/pull/617/files/9ea652d2bc590f081f954c9974aff6171111ef74#r69787231), @adamralph. :stuck_out_tongue_winking_eye:

**Comment by blairconrad** [comment]
Oh, we need **breaking** in addition to **removal**? I thought the latter implied the former. Still, best to be explicit. Thanks for the label!

**Comment by adamralph** [comment]
Ah yes, I was sure you'd done a similar search but I couldn't find it.

**Comment by blairconrad** [comment]
The trick was to search for service-ghost. Which would be hard to do before doing the research...

**Comment by adamralph** [comment]
We do have some removals not marked as breaking: https://github.com/FakeItEasy/FakeItEasy/issues?utf8=%E2%9C%93&q=is%3Aclosed%20label%3Aremoval%20-label%3Abreaking

**Comment by blairconrad** [comment]
So we do! Of course, I would've called all of those (save maybe the logging one) breaking.

**Comment by thomaslevesque** [comment]
`removal` does't imply `breaking`, if we're removing something that doesn't affect the public API. So I think it makes sense to have both.

**Comment by adamralph** [comment]
And of course, we also have many breaking changes which are not removals. I think it's nice to be able to filter on the `breaking` label to see all breaking changes, rather than have to combine it with a query for `removal`.

Regarding those [removals not marked as breaking](https://github.com/FakeItEasy/FakeItEasy/issues?utf8=%E2%9C%93&q=is%3Aclosed%20label%3Aremoval%20-label%3Abreaking) the only controversial ones (I think) are the removals of support for SL3, SL5, Win8, Win8.1 and .NET 3.5. I remember that we came to the conclusion that NuGet will simply not upgrade you when targeting those platforms, so there's no way to get "broken".

**Comment by blairconrad** [comment]
>  if we're removing something that doesn't affect the public API

… I'd call it refactoring or tech-debt.

Regardless, I wasn't trying to start a thing. I was genuinely happy to see the breaking tag added.

So, I think @thomaslevesque thumb-upped the issue. @adamralph?

**Comment by adamralph** [comment]
I'm :+1: for removing it too. The feature isn't even documented.

I'd be interested to know why @patrik-hagne added it in the first place though. Perhaps there was once a compelling use case?

**Comment by adamralph** [comment]
It goes all the way back to the [initial commit](https://github.com/FakeItEasy/FakeItEasy/tree/8461f5fca6eec74c6ede6dbe7562a6137688a1af/Source/FakeItEasy/SelfInitializedFakes) so we can't tease anything out of the history.

**Comment by blairconrad** [comment]
Oh! Good idea searching back.

**Comment by thomaslevesque** [comment]
Since we all agree, I changed the tags to `ready`.

Actually, I already have a branch ready with the changes. Should it be merged on a new `vnext` branch, or on `master`? If it's on `master`, new work for 2.x will have to be done on a maintenance branch.

**Comment by adamralph** [comment]
:-1: to branching. IMHO, if we're going to do this now, master should become 3.0 and 2.x should go into maintenance mode (bug fixes only). That means no new feature releases before 3.0, and I think .NET core support should be part of 3.0 since we may find that we need more breaking changes to support it.

If we go down this path, we should release 2.2 before making master 3.0.

**Comment by adamralph** [comment]
Also, this could be premature. It depends how soon we want to merge in the .NET core support.

**Comment by thomaslevesque** [comment]
> IMHO, if we're going to do this now, master should become 3.0 and 2.x should go into maintenance mode (bug fixes only). That means no new feature releases before 3.0

Can't we continue adding (non-breaking) features to 2.x while we start working on 3.0? I mean, it might be a while before we release 3.0... we don't want 2.x to end up like 1.x, which received no new features for a very long time.

**Comment by adamralph** [comment]
If it's a long time before we release 3.0, then adding features to 2.x in the meantime would maintaining and merging between two branches for a long time and that is something I want to avoid.

This is why I think we should hold off for a while until we've better idea of how far off we are with .NET core. We don't have any other breaking changes saved up for 3.0, so we can release 3.0 straight away as soon as .NET core gets merged in. Also, I'm guessing that these removal is fairly simple to do, so it could even just be a commit in the .NET Core PR branch.

**Comment by thomaslevesque** [comment]
> We don't have any other breaking changes saved up for 3.0, so we can release 3.0 straight away as soon as .NET core gets merged in.

If we're going to release a new major version with breaking changes, I think we might as well make another pass on the public API (easy now that we have API approval tests) and remove anything that shouldn't be there.

**Comment by blairconrad** [comment]
I'd definitely release 2.2 (or 2.x if we end up with more stuff) before 3.0. Heck, we could release 2.2 any time, as far as I'm concerned.

While I don't hate the idea of having a vNext branch around for a little while (perhaps rebasing as we go, instead of merging), I'm okay not doing it for a while, either. (Also, individuals who really like the idea of a vNext branch could just keep their own. It's pretty much what we're doing with jeremymeng's coreclr.)

I'm all for another pass on the public API before a major release as well.

**Comment by morgul77** [comment]
Hi

I was just spiking something for our intetgration testing. 

My goal is to know when a method is called, deep in the stack of an object tree. We use autofac as DI.

So the code I came up with is like this:
<pre> builder.RegisterType<DirectoryContentService>()
                    .As<IDirectoryContentService>()
                    .OnActivating(b =>
                    {
                        wrappedDirectoryService = A.Fake<DirectoryContentService>(y => y.Wrapping(b.Instance)
                                                                                        .RecordedBy(
                                    A.Fake<ISelfInitializingFakeRecorder>(m => m.ConfigureFake(n =>
                                       {
                                           A.CallTo(() => n.IsRecording).Returns(true);
                                           A.CallTo(
                                               () =>
                                                   n.RecordCall(
                                                       A<ICompletedFakeObjectCall>.That.Matches(o => o.Method.Name == "GetCachedContents" && ((List<DirectoryContent>)o.ReturnValue).Count == 1))).Invokes(
                                                           j =>
                                                           {
                                                               tcs.TrySetResult(null);
                                                           });
                                       }))).Implements<IDirectoryContentService>());

                        b.ReplaceInstance(wrappedDirectoryService);
</pre>


Now; the one of the few things I found on ISelfInitializingFakeRecorder was here, and was actually going to ask if there was a nicer way of doing this (it looks quite ugly and inflexible, refinement is necessary), only to find that it might be gone?

With integration testing and autofac, this seemed like nice approach, instead of waiting and relying on sideeffects with no real hooks.

If this is the possibility you are thinking for removing, I vote no:)

If there's a better approach, please, enlighten me.

**Comment by thomaslevesque** [comment]
@morgul77, if you just want to intercept calls, there's a much simpler way:

```
var fakeManager = Fake.GetManager(theFake);
var listener = new MyListener();
fake.AddInterceptionListener(listener);
```

MyListener looks like this:

```
class MyListener : IInterceptionListener
{
    public void OnBeforeCallIntercepted(IFakeObjectCall call)
    {
    }

    public void OnAfterCallIntercepted(ICompletedFakeObjectCall call, IFakeObjectCallRule rule)
    {
    }
}
```

**Comment by morgul77** [comment]
Ok, that works too. So that's good. What I'm hoping for is a simple syntax, preferably a oneliner, that will do the trick. That is, no creating classes, and a simpler syntax (but this is now going off topic).

Some syntax like:
<pre>
 A.Fake<DirectoryContentService>(y => y.Wrapping(b.Instance).Implements<IDirectoryContentService>().ConfigureFake(
                            n =>
                            {
                                A.CallTo(() => n.GetCachedContents()).WhenReturnValueMatches(x => x.Count == 1, x => tcs.TrySetResult(x))


                            }));

</pre>

**Comment by adamralph** [comment]
What the latest thinking around this? Are we going to keep it in net40 and leave it out of netstandard?

**Comment by adamralph** [comment]
Marking as P1 since a decision here will dictate whether our netstandard release will be major or minor.

**Comment by blairconrad** [comment]
I say drop it completely.

**Comment by adamralph** [comment]
I'm on the fence.

**Comment by thomaslevesque** [comment]
+1 for dropping it, because I think it has little value.

That being said, we *could* keep it, disable it in netstandard, and re-enable it when the serialization APIs are reintroduced (next year according to the .NET Core roadmap). But I don't really think it's worth the trouble.

> I'm on the fence.

What makes you hesitate?

**Comment by blairconrad** [comment]
> What makes you hesitate?

Is it [this comment](https://github.com/FakeItEasy/FakeItEasy/pull/617#discussion_r70092545)?

**Comment by adamralph** [comment]
If we didn't have to support .net core now, and we were just continuing with the net40 TFN for the foreseeable future, would we be considering dropping it in the next major?

If the answer is no, then I believe we shouldn't drop it now just to cater for .net core temporarily not having binary serialisation.

If the answer is yes, I believe we shouldn't be restricting ourselves to just this feature. We should do a wider analysis and come up with a complete list of features we want to drop when we release 3.0.

**Comment by thomaslevesque** [comment]
> If we didn't have to support .net core now, and we were just continuing with the net40 TFN for the foreseeable future, would we be considering dropping it in the next major?

Assuming it came to our attention, I'd say yes, but perhaps I'm biased because I just don't like the feature. Before the work on .NET Core started, I don't think any of us really knew what self initialized fakes were. We were basically ignoring them, so we didn't consider removing them...

**Comment by blairconrad** [comment]
Well, @thomaslevesque had proposed reviewing the public API as part of the next major release (which I support), so it's a possibility.
Seeing the amount of code devoted to the feature, and considering the shortcomings we've discussed in the feature (arcane API, the necessity of _exactly_ the same calls coming in when replaying, the fact that the calls' arguments aren't examined), I'd be tempted. Sort of like @thomaslevesque, I don't like the feature. Partly for the reasons I just mentioned. And a little bit because I don't think that it buys users that much over just writing a couple of `A.CallTo`s. And there are other downsides when compared to the `A.CallTo`s. The self-initialized fake "script" is stored outside of the code, so uses of it would make tests harder to understand. And (I assume) the binary serialized version of the script would be unreadable by humans. We've made the comment in the past that FIE is an opinionated testing framework, so I guess by advocating the removal of the self-initialized fakes, voting against this style of testing.

>  Before the work on .NET Core started, I don't think any of us really knew what self initialized fakes were. We were basically ignoring them, so we didn't consider removing them...

Nah, we knew what they were. There'd been some work early in 2.0 where I'd beefed up the specs before working on some other features. Perhaps it's more accurate to say that _I_ knew what they were. Still, they were easy to ignore, and I hadn't looked that deeply into their limitations.

**Comment by adamralph** [comment]
OK, I agree that we should remove the feature.

In that case, given that we don't have any non-breaking features/bug fixes in progress, it seems our next release will be 3.0. I also just noticed that https://github.com/FakeItEasy/FakeItEasy/issues/803 is a breaking change, so this will also allow that to go in. The only other breaking change we have raised now is https://github.com/FakeItEasy/FakeItEasy/issues/803 but we haven't yet reached consensus as to what to do there.

I do think we should scour the rest of the API to see what else we would like to remove now.

**Comment by blairconrad** [comment]
> I do think we should scour the rest of the API to see what else we would like to remove now.

Idea: start a PR with a changed [ApiApproval.ApproveApi.approved.txt](https://github.com/FakeItEasy/FakeItEasy/blob/master/tests/FakeItEasy.Tests.Approval/ApiApproval.ApproveApi.approved.txt) for discussion.

**Comment by blairconrad** [comment]
I'm leery of starting down the 3.0 path yet, since I don't think we still know how long it'll take to get a .NET Standard build. Castle.Core still hasn't gone "production release".
I don't want to be in the position where we _can't_ publish a minor release for 8 months…

**Comment by adamralph** [comment]
In that case, let's not merge anything for 3.0 into master just yet.

I know what you're going to say now... "let's start a 3.0 branch" :smile: If we do indeed want to do that, i.e. start developing the next major whilst adding features to the current major then I think we need to switch to gitflow instead of githubflow. We can talk more about that on gitter.

**Comment by adamralph** [comment]
Bumping back to P3. Not at all important when we do this, nor in which milestone.

**Comment by blairconrad** [comment]
I wonder if we shouldn't deprecate the feature to signal a lack of support and eventual removal.

**Comment by adamralph** [comment]
Yes! Great idea!

**Comment by adamralph** [comment]
Spin off an issue labelled `deprecation`?

**Comment by blairconrad** [comment]
> Spin off an issue labelled `deprecation`?

As you wish, Buttercup.

**Comment by patrik-hagne** [comment]
Have a look here if you want the original commits...
https://fakes.codeplex.com

We use this feature for some integration tests where we stub out 3d party
web services, record it once, then the tests are fast and deterministic.

Den 3 aug 2016 20:11 skrev "Blair Conrad" <notifications@github.com>:

**Comment by adamralph** [comment]
Woah, "Legend Fakes"! A piece of history I didn't know about...

**Comment by patrik-hagne** [comment]
Your pretty much the second person to know about it! ☺

Den 3 aug 2016 23:52 skrev "Adam Ralph" <notifications@github.com>:

**Comment by blairconrad** [comment]
@patrik-hagne said

> We use this feature for some integration tests where we stub out 3d party
> web services, record it once, then the tests are fast and deterministic.

When I first read this, I thought that maybe you meant that during your integration test, you ran against the real service once, and that the rest of the tests during the run were fast, which was a new usage to me.
Or did you mean that you persist the results of the first fun and then you have fast, deterministic tests for weeks to come?

If the latter, do you essentially create a single fake for the service and record the interaction over the entire integration test suite, so that the whole suite follows the script on subsequent re-runs, or do you record interactions for what is essentially one test, and then recreate the fake (with the same script) for each test?

**Comment by adamralph** [comment]
Moving back to the backlog, since it seems we're not yet 100% sure that we want to do this.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.0.0).
